embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+    /// currently full and sending would require blocking.
+    Full(T),
+
+    /// The receive half of the channel was explicitly closed or has been
+    /// dropped.
+    Closed(T),
+}
+
+impl<T> fmt::Display for TrySendError<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "{}",
+            match self {
+                TrySendError::Full(..) => "no available capacity",
+                TrySendError::Closed(..) => "channel closed",
+            }
+        )
+    }
+}
+
+struct ChannelState<T, const N: usize> {
+    buf: [MaybeUninit<UnsafeCell<T>>; N],
+    read_pos: usize,
+    write_pos: usize,
+    full: bool,
+    closing: bool,
+    closed: bool,
+    receiver_registered: bool,
+    senders_registered: u32,
+    receiver_waker: Option<Waker>,
+    senders_waker: Option<Waker>,
+}
+
+impl<T, const N: usize> ChannelState<T, N> {
+    const INIT: MaybeUninit<UnsafeCell<T>> = MaybeUninit::uninit();
+
+    const fn new() -> Self {
+        let buf = [Self::INIT; N];
+        let read_pos = 0;
+        let write_pos = 0;
+        let full = false;
+        let closing = false;
+        let closed = false;
+        let receiver_registered = false;
+        let senders_registered = 0;
+        let receiver_waker = None;
+        let senders_waker = None;
+        ChannelState {
+            buf,
+            read_pos,
+            write_pos,
+            full,
+            closing,
+            closed,
+            receiver_registered,
+            senders_registered,
+            receiver_waker,
+            senders_waker,
+        }
+    }
+}
+
+/// A a bounded mpsc channel for communicating between asynchronous tasks
+/// with backpressure.
+///
+/// The channel will buffer up to the provided number of messages.  Once the
+/// buffer is full, attempts to `send` new messages will wait until a message is
+/// received from the channel.
+///
+/// All data sent will become available in the same order as it was sent.
+pub struct Channel<M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    mutex: M,
+    state: ChannelState<T, N>,
+}
+
+pub type WithCriticalSections = CriticalSectionMutex<()>;
+
+impl<T, const N: usize> Channel<WithCriticalSections, T, N> {
+    /// Establish a new bounded channel using critical sections. Critical sections
+    /// should be used only single core targets where communication is required
+    /// from exception mode e.g. interrupt handlers. To create one:
+    ///
+    /// ```
+    /// use embassy::util::mpsc;
+    /// use embassy::util::mpsc::{Channel, WithCriticalSections};
+    ///
+    /// // Declare a bounded channel of 3 u32s.
+    /// let mut channel = Channel::<WithCriticalSections, u32, 3>::with_critical_sections();
+    /// // once we have a channel, obtain its sender and receiver
+    /// let (sender, receiver) = mpsc::split(&channel);
+    /// ```
+    pub const fn with_critical_sections() -> Self {
```

Comment:
Would it be possible to have just one `new()` constructor? The channel already knows the type of mutex it should use from the generic params.

Reply:
I wanted to do that too. Here's what I get though:

```
error[E0034]: multiple applicable items in scope
   --> embassy/src/util/mpsc.rs:779:78
    |
779 |         static mut CHANNEL: Channel<WithCriticalSections, u32, 3> = Channel::new();
    |                                                                              ^^^ multiple `new` found
    |
note: candidate #1 is defined in an impl for the type `mpsc::Channel<mutex::CriticalSectionMutex<()>, T, N>`
   --> embassy/src/util/mpsc.rs:453:5
    |
453 |     pub const fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #2 is defined in an impl for the type `mpsc::Channel<mutex::ThreadModeMutex<()>, T, N>`
   --> embassy/src/util/mpsc.rs:477:5
    |
477 |     pub const fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
note: candidate #3 is defined in an impl for the type `mpsc::Channel<mutex::NoopMutex<()>, T, N>`
   --> embassy/src/util/mpsc.rs:498:5
    |
498 |     pub const fn new() -> Self {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
`lock_api`'s  `RawMutex` solves this with the [`const INIT`](https://docs.rs/lock_api/0.4.4/lock_api/trait.RawMutex.html#associatedconstant.INIT). WIth it, `new()` can generically construct a mutex instance without knowing which type it is.

Reply:
I tried a few things here. I'm likely missing some foo. Would you be able to provide a bit more of an example?

Reply:
Commit bd6cd3471d888e7d72a132f13b53aabf09414586 attends to this. Thanks for the code @Dirbaio!